crypto/fipsmodule/sha/sha3.c

```diff
@@ -196,36 +180,73 @@ int SHA3_Update(KECCAK1600_CTX *ctx, const void *data, size_t len) {
   return 1;
 }
 
-int SHA3_Final(uint8_t *md, KECCAK1600_CTX *ctx) {
+// FIPS202_Finalize processes padding and absorb of last input block
+// This function should be called once to finalize absorb and initiate squeeze phase
+static int FIPS202_Finalize(uint8_t *md, KECCAK1600_CTX *ctx) {
   size_t block_size = ctx->block_size;
   size_t num = ctx->buf_load;
 
-  if (ctx->md_size == 0) {
+  if (ctx->state == KECCAK1600_STATE_SQUEEZE || 
+      ctx->state == KECCAK1600_STATE_FINAL ) {
+    return 0;
+  }
+
+  // Pad the data with 10*1. Note that |num| can be |block_size - 1|
+  // in which case both byte operations below are performed on
+  // the same byte.
+  memset(ctx->buf + num, 0, block_size - num);
+  ctx->buf[num] = ctx->pad;
+  ctx->buf[block_size - 1] |= 0x80;
+
+  if (Keccak1600_Absorb(ctx->A, ctx->buf, block_size, block_size) != 0) {
+    return 0;
+  }
+
+  return 1;
+}
+
+// SHA3 APIs implement SHA3 functionalities on top of FIPS202 API layer
+int SHA3_Init(KECCAK1600_CTX *ctx, size_t bit_len) {
+  if (bit_len == SHA3_224_DIGEST_BITLENGTH || 
+      bit_len == SHA3_256_DIGEST_BITLENGTH || 
+      bit_len == SHA3_384_DIGEST_BITLENGTH || 
+      bit_len == SHA3_512_DIGEST_BITLENGTH) { 
+        // |block_size| depends on the SHA3 |bit_len| output (digest) length
+        return FIPS202_Init(ctx, SHA3_PAD_CHAR, SHA3_BLOCKSIZE(bit_len), bit_len);
+  }
+  return 0;
+}
+
+int SHA3_Update(KECCAK1600_CTX *ctx, const void *data, size_t len) {
+  if (ctx == NULL) {
+    return 0;
+  }
+
+  if (len == 0) {
     return 1;
   }
 
-  if (ctx->padded == 0) {
-    // Pad the data with 10*1. Note that |num| can be |block_size - 1|
-    // in which case both byte operations below are performed on
-    // the same byte.
-    memset(ctx->buf + num, 0, block_size - num);
-    ctx->buf[num] = ctx->pad;
-    ctx->buf[block_size - 1] |= 0x80;
+  return FIPS202_Update(ctx, data, len);
+}
 
-    if (Keccak1600_Absorb(ctx->A, ctx->buf, block_size, block_size) != 0) {
-      return 0;
-    }
+// SHA3_Final should be called once to process final digest value
+// |ctx->state| flag does not need to be updated
+int SHA3_Final(uint8_t *md, KECCAK1600_CTX *ctx) {
```

Comment:
~Should there be a check that it was called in the right state?~ It's done in `FIPS202_Finalize()`